### PR TITLE
RLP Refactor

### PR DESCRIPTION
**Pull request by AlanVerbner**
## Description

The following changes were made to RLP code:

- RLPValue and RLPList traits replaced by case classes (breaking)
- Match cases replaced by if statements if no unboxing
- all exceptions replaced by RuntimeExceptions
- ByteString repalced by Array[Byte]
- Prevent Long values to be decoded if they have more than 8 Bytes (breaking - low impact)
- Tests and Scoverage plugin added
- RLPEncoding package object created and code moved into it (breaking)
- RPL decoding of an empty array now results in an exception (breaking - low impact)

### CODE REVIEW

**Comment by mcsherrylabs** [comment]
Hi Alan, I made a few changes, just trying to get the API into the package object. Please have a look, and if I haven't broken anything, LGTM...

**Comment by AlanVerbner** [comment]
@mcsherrylabs reviewing now

**Comment by AlanVerbner** [comment]
@mcsherrylabs LGTM thanks! :smile:

File: src/main/scala/io/iohk/ethereum/rlp/package.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+  trait RLPEncoder[T] {
+    def encode(obj: T): RLPEncodeable
+  }
+
+  trait RLPDecoder[T] {
+    def decode(rlp: RLPEncodeable): T
+  }
+
+  def encode[T](input: T)(implicit enc: RLPEncoder[T]): Array[Byte] = RLP.encode(enc.encode(input))
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/rlp/package.scala]: why do we need these methods? Can't we just have `def encode[T](input: T)(implicit enc: RLPEncoder[T])` in `RLP` obj?

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/rlp/package.scala]: @LukasGasior1 the goal is to have all public RLP api in `rlp` package object, does that make sense?

Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/rlp/package.scala]: I don't know why that's the goal, but it's not a big deal anyway

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/rlp/package.scala]: @mcsherrylabs :arrow_up:

Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/rlp/package.scala]: @lukasGasior1, in this case the API for RLP is small enough to be declared in the package class. That helps clients of the module understand what's available at a glance. When clients import the package they have all the functionality available bar the supporting implicits.

**Pull request merged**